File: sources/base/ui/src/main/java/com/egoriku/ui/arch/pvm/BaseContract.java
package com.egoriku.ui.arch.pvm;

import android.arch.lifecycle.Lifecycle;

public interface BaseContract {

    interface View {
    }

    interface Presenter<V extends BaseContract.View> {

        void attachLifecycle(Lifecycle lifecycle);
        void detachLifecycle(Lifecycle lifecycle);

        void attachView(V view);
        void detachView();

        V getView();

        boolean isViewAttached();

        void onPresenterCreated();
        void onPresenterDestroy();
    }
}

File: sources/base/ui/src/main/java/com/egoriku/ui/arch/pvm/BaseContract.kt
package com.egoriku.ui.arch.pvm

import android.arch.lifecycle.Lifecycle

interface BaseContract {

    interface View

    interface Presenter<V : BaseContract.View> {

        val view: V?

        fun attachLifecycle(lifecycle: Lifecycle)
        fun detachLifecycle(lifecycle: Lifecycle)

        fun attachView(view: V)
        fun detachView()

        fun onPresenterCreated()
        fun onPresenterDestroy()
    }
}
